pgcli/packages/pgspecial/main.py

```diff
@@ -12,12 +12,81 @@
 SpecialCommand = namedtuple('SpecialCommand',
         ['handler', 'syntax', 'description', 'arg_type', 'hidden', 'case_sensitive'])
 
-COMMANDS = {}
-
 @export
 class CommandNotFound(Exception):
     pass
 
+
+@export
+class PGSpecial(object):
+
+    # Default static commands that don't rely on PGSpecial state are registered
+    # via the special_command decorator and stored in default_commands
+    default_commands = {}
+
+    def __init__(self):
+        self.timing_enabled = True
+
+        self.commands = self.default_commands.copy()
+
+        self.timing_enabled = False
+        self.expanded_output = False
+
+        self.register(self.show_help, '\\?', '\\?', 'Show Help.',
+                      arg_type=NO_QUERY)
+
+        self.register(self.toggle_expanded_output, '\\x', '\\x',
+                      'Toggle expanded output.', arg_type=NO_QUERY)
+
+        self.register(self.toggle_timing, '\\timing', '\\timing',
+                      'Toggle timing of commands.', arg_type=NO_QUERY)
+
+    def register(self, *args, **kwargs):
+        register_special_command(*args, command_dict=self.commands, **kwargs)
+
+    def execute(self, cur, sql):
+        commands = self.commands
+        command, verbose, pattern = parse_special_command(sql)
+
+        if (command not in commands) and (command.lower() not in commands):
+            raise CommandNotFound
+
+        try:
+            special_cmd = commands[command]
+        except KeyError:
+            special_cmd = commands[command.lower()]
+            if special_cmd.case_sensitive:
+                raise CommandNotFound('Command not found: %s' % command)
+
+        if special_cmd.arg_type == NO_QUERY:
+            return special_cmd.handler()
+        elif special_cmd.arg_type == PARSED_QUERY:
+            return special_cmd.handler(cur=cur, pattern=pattern, verbose=verbose)
+        elif special_cmd.arg_type == RAW_QUERY:
+            return special_cmd.handler(cur=cur, query=sql)
+
+    def show_help(self):
+        headers = ['Command', 'Description']
+        result = []
+
+        for _, value in sorted(self.commands.items()):
+            if not value.hidden:
+                result.append((value.syntax, value.description))
+        return [(None, result, headers, None)]
+
+    def toggle_expanded_output(self):
+        self.expanded_output = not self.expanded_output
+        message = u"Expanded display is "
+        message += u"on." if self.expanded_output else u"off."
+        return [(None, None, None, message)]
+
+    def toggle_timing(self):
```

Comment:
Good call to make `toggle_timing`, `toggle_expanded_output` and `show_help` to be methods on the PGSpecial class. :)